cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -132,6 +133,9 @@ struct ARROW_EXPORT FileSelector {
   bool recursive;
   /// The maximum number of subdirectories to recurse into.
   int32_t max_recursion;
+  /// How many partitions should be processed in parallel. May not be supported by all
+  /// implementations of `GetFileSystemGenerator`.
+  util::optional<int32_t> partitions_readahead;
```

Comment:
Are there any practical reasons why you do avoid optionals in public APIs? Does Arrow try to maintain strict ABI stability across multiple releases? If not, then I guess it should be completely fine.

Regarding `directory_readahead`: I agree, the term `partitions` does not belong to this level of abstraction. I'll change it to `directory_readahead`, then.

Reply:
We don't have any ABI guarantees. But when there is a well-known default value, it doesn't make sense to pass an optional, IMHO.

Reply:
However, I also don't understand why this is an attribute of `FileSelector`. This sounds more like an implementation-specific know that should probably be in `LocalFileSystemOptions`.

Reply:
(I'm also curious why this needs to be exposed at all)

Reply:
>We don't have any ABI guarantees. But when there is a well-known default value, it doesn't make sense to pass an optional, IMHO. 

Maybe you are right, this can be supplied a reasonable default value, so no need to make it optional.

> However, I also don't understand why this is an attribute of FileSelector. This sounds more like an implementation-specific know that should probably be in LocalFileSystemOptions.

Well, I don't have a strong opinion on this one. Don't see anything wrong about it being a part of FileSelector (which is exactly designed to specify details of file selection algorithms). Though, since this option is only applicable for `LocalFileSystem`, maybe it makes sense to hide it behind `LocalFileSystemOptions` thing.

> (I'm also curious why this needs to be exposed at all)

To be able to fine-tune the behavior if the default doesn't work well in a particular case, or better performance can be achieved with another value. For example, various filesystems have varying capabilities regarding parallel IO, e.g. XFS, which is the only FS I know of, that is capable of truly async IO.